pkg/notify/notify.go

```diff
@@ -3,166 +3,218 @@ package notify
 import (
 	"bytes"
 	"fmt"
-	"github.com/go-kit/kit/log"
-	"github.com/go-kit/kit/log/level"
 	"io/ioutil"
 	"reflect"
 	"strings"
 	"time"
 
-	"github.com/free/jiralert/pkg/config"
-	"github.com/free/jiralert/pkg/template"
-
 	"github.com/andygrunwald/go-jira"
 	"github.com/free/jiralert/pkg/alertmanager"
+	"github.com/free/jiralert/pkg/config"
+	"github.com/free/jiralert/pkg/template"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
 	"github.com/trivago/tgo/tcontainer"
 )
 
-// Receiver wraps a JIRA client corresponding to a specific Alertmanager receiver, with its configuration and templates.
+// TODO(bwplotka): Consider renaming this package to ticketer.
+
+type jiraIssueService interface {
+	Search(jql string, options *jira.SearchOptions) ([]jira.Issue, *jira.Response, error)
+	GetTransitions(id string) ([]jira.Transition, *jira.Response, error)
+
+	Create(issue *jira.Issue) (*jira.Issue, *jira.Response, error)
+	UpdateWithOptions(issue *jira.Issue, opts *jira.UpdateQueryOptions) (*jira.Issue, *jira.Response, error)
+	DoTransition(ticketID, transitionID string) (*jira.Response, error)
+}
+
+// Receiver wraps a specific Alertmanager receiver, with its configuration and templates to notify given ticket service.
 type Receiver struct {
-	conf   *config.ReceiverConfig
-	tmpl   *template.Template
-	client *jira.Client
+	logger log.Logger
+	client jiraIssueService
+	// TODO(bwplotka): Consider splitting receiver config with ticket service details.
+	conf *config.ReceiverConfig
+	tmpl *template.Template
+
+	timeNow func() time.Time
 }
 
-// NewReceiver creates a Receiver using the provided configuration and template.
-func NewReceiver(c *config.ReceiverConfig, t *template.Template) (*Receiver, error) {
-	tp := jira.BasicAuthTransport{
-		Username: c.User,
-		Password: string(c.Password),
-	}
-	client, err := jira.NewClient(tp.Client(), c.APIURL)
+// NewReceiverWithClient creates a Receiver using the provided configuration, template and ticketClientService.
+func NewReceiverWithClient(logger log.Logger, c *config.ReceiverConfig, t *template.Template, client jiraIssueService) *Receiver {
+	return &Receiver{logger: logger, conf: c, tmpl: t, client: client, timeNow: time.Now}
+}
+
+// Notify translates alert to a ticket.
+func (r *Receiver) Notify(data *alertmanager.Data) (bool, error) {
+	project, err := r.tmpl.Execute(r.conf.Project, data)
 	if err != nil {
-		return nil, err
+		return false, errors.Wrap(err, "render project")
```

Comment:
`s/render/generate/`? `generating`? Seems to be the term used everywhere else (JIRAlert documentation and Golang template library).